lib/kazan/watcher.ex

```diff
@@ -0,0 +1,164 @@
+  and a pid to which to send received messages.  The initial *resource_version* can be
+  obtained by doing a normal, non-watch request to the same endpoint.
+  ```
+  Kazan.Watcher.start_link(request, resource_version: rv, send_to: self())
+  ```
+  3. In your client code you receive messages for each `%Versioned.Event{}`.
+    You can pattern match on the object type if you have multiple watchers configured.
+    For example, if the client is a `GenServer`
+  ```
+    # type is "ADDED", "DELETED" or "MODIFIED"
+    def handle_info(%Kazan.Models.Versioned.Event{object: object, type: type}, state) do
+      case object do
+        %Kazan.Models.V1.Namespace{} = namespace ->
+          process_namespace_event(type, namespace)
+
+        %Kazan.Models.V1.Job{} = job ->
+          process_job_event(type, job)
+      end
+      {:noreply, state}
+    end
+  ```
+  """
+
+  use GenServer
+  alias Kazan.LineBuffer
+  require Logger
+
+  defmodule State do
+    @moduledoc """
+    Stores the internal state of the watcher.
+
+    Includes:
+
+    resource_version:
+
+    The K8S *resource_version* (RV) is used to version each change in the cluster.
+    When we listen for events we need to specify the RV to start listening from.
+    For each event received we also receive a new RV, which we store in state.
+    When the watch eventually times-out (it is an HTTP request with a chunked response),
+    we can then create a new request passing the latest RV that was received.
+
+    buffer:
+
+    Chunked HTTP responses are not always a complete line, so we must buffer them
+    until we have a complete line before parsing.
+    """
+    defstruct id: nil, request: nil, send_to: nil, buffer: LineBuffer.new(), rv: nil
+  end
+
+  @doc "Starts an async request send_to is the process to send messages to; resource_version can also be passed"
+  def start_link(%Kazan.Request{} = request, opts) do
+    send_to = Keyword.fetch!(opts, :send_to)
+    rv = Keyword.fetch!(opts, :resource_version)
+    GenServer.start_link(__MODULE__, [request, rv, send_to])
+  end
+
+  @impl GenServer
+  def init([request, rv, send_to]) do
+    Logger.info "Watcher init pid: #{inspect self()} rv: #{rv} request: #{inspect request}"
+    {:ok, id} = Kazan.Client.run(request, stream_to: self())
+    {:ok, %State{id: id, request: request, rv: rv, send_to: send_to}}
+  end
+
+  @impl GenServer
+  def handle_info(%HTTPoison.AsyncStatus{code: 200}, state) do
+    {:noreply, state}
+  end
+
+  @impl GenServer
+  def handle_info(%HTTPoison.AsyncHeaders{}, state) do
+    {:noreply, state}
+  end
+
+  @impl GenServer
+  def handle_info(%HTTPoison.AsyncChunk{chunk: chunk, id: request_id},
+                  %State{id: id, buffer: buffer, rv: rv, send_to: send_to} = state) do
+    if request_id != id do
+      Logger.warn "Received chunk for wrong request_id: #{inspect request_id} expect: #{inspect id}"
+      {:noreply, state}
```

Comment:
This shouldn't really happen, so I'm wondering if the correct thing to do would be to let it crash so supervisor can restart things in a hopefully-good state.

Reply:
Yes, I agree.